src/EnvelopeCreator/EnvelopeCreator.php

```diff
@@ -0,0 +1,130 @@
+    ) {
+        $this->router = $router;
+        $this->logger = $logger;
+        $this->stopwatch = $stopwatch;
+        $this->createDocument = $createDocument;
+        $this->createRecipient = $createRecipient;
+        $this->defineEnvelope = $defineEnvelope;
+        $this->sendEnvelope = $sendEnvelope;
+    }
+
+    /**
+     * @throws FileNotFoundException
+     *
+     * @return string path to redirect
+     *
+     * @throw InvalidArgumentException
+     */
+    public function createEnvelope(EnvelopeBuilder $envelopeBuilder): string
+    {
+        try {
+            $this->validate($envelopeBuilder);
+
+            $key = '[Docusign] Create document';
+            $this->stopwatch->start($key);
+            $this->createDocument->handle($envelopeBuilder);
+            $this->stopwatch->stop($key);
+
+            $this->defineEnvelope->handle($envelopeBuilder);
+
+            $key = '[Docusign] Send envelope';
+            $this->stopwatch->start($key);
+            $this->sendEnvelope->handle($envelopeBuilder);
+            $this->stopwatch->stop($key);
+
+            if (EnvelopeBuilder::MODE_REMOTE === $envelopeBuilder->mode) {
+                return $this->getCallbackRoute($envelopeBuilder);
+            }
+
+            $key = '[Docusign] Create recipient';
+            $this->stopwatch->start($key);
+            $viewUrl = $this->createRecipient->handle($envelopeBuilder);
+            $this->stopwatch->stop($key);
+
+            return $viewUrl->getUrl();
+        } catch (ApiException $exception) {
+            $this->logger->critical('Unable to send a document to DocuSign.', [
+                'document' => $envelopeBuilder->document,
+                'signers' => $envelopeBuilder->signers,
+                'envelope' => $envelopeBuilder->envelopeDefinition,
+                'request' => $exception->getResponseBody(),
+            ]);
+            if (!empty($key)) {
+                $this->stopwatch->stop($key);
+            }
+
+            throw new UnableToSignException($exception->getMessage());
+        } finally {
+            $this->reset($envelopeBuilder);
+        }
+    }
+
+    private function validate(EnvelopeBuilder $envelopeBuilder): void
+    {
+        Assert::notNull($envelopeBuilder->filePath);
+        Assert::notEmpty($envelopeBuilder->docReference);
+    }
+
+    /**
+     * Returns the callback.
+     */
+    private function getCallbackRoute(EnvelopeBuilder $envelopeBuilder): string
+    {
+        try {
+            Assert::regex($envelopeBuilder->callbackRouteName, '#^https?://.+(:[0-9]+)?$#');
+
+            return $envelopeBuilder->callbackRouteName;
+        } catch (InvalidArgumentException $exception) {
+            return $this->router->generate($envelopeBuilder->callbackRouteName, array_unique(['envelopeId' => $envelopeBuilder->envelopeId] + $envelopeBuilder->callbackParameters), Router::ABSOLUTE_URL);
+        }
+    }
+
+    public function reset(EnvelopeBuilder $envelopeBuilder): void
```

Comment:
I think this reset method should be present on the EnvelopeBuilder class, implementing the [`ResetInterface`](https://github.com/symfony/contracts/blob/master/Service/ResetInterface.php)